syntax = "proto3";

package minecraft;

service MinecraftEnv {
  // 健康检查
  rpc Health (Empty) returns (HealthResp);

  // 批量创建
  rpc BatchCreate (BatchCreateReq) returns (BatchCreateResp);

  // 查询环境状态
  rpc EnvStatus (EnvIdReq) returns (EnvStatusResp);

  // 同步 reset
  rpc Reset (FastResetReq) returns (ResetResp);

  // 同步 step（调试用）
  rpc Step (StepReq) returns (StepResp);

  // 异步 step：提交
  rpc SubmitStep (StepReq) returns (StepResultResp);

  // 异步 step：轮询结果
  rpc GetStepResult (EnvIdReq) returns (StepResultResp);

  // 关闭环境
  rpc CloseEnv (EnvIdReq) returns (CloseResp);

  // 获取异步 reset 结果
  rpc GetResetResult (GetResetResultReq) returns (ResetResp);

}

message Empty {}

message HealthResp {
  string status = 1;
  bool ray_initialized = 2;
  int32 num_environments = 3;
}

message BatchCreateReq {
  int32 count = 1;
  string env_name = 2;
  repeated string env_kwargs = 3;   // 每段 JSON 序列化成 string
}

message BatchCreateResp {
  repeated string env_ids = 1;
}

message EnvIdReq {
  string env_id = 1;
}

message FastResetReq {
  string env_id = 1;
  string config = 2;
}

message EnvStatusResp {
  string env_id = 1;
  string status = 2;   // idle / busy / failed / not_found
  bool ready = 3;
}

message ResetResp {
  string observation_json = 1; // 序列化后的 dict
  string info_json = 2;
}

message StepReq {
  string env_id = 1;
  string action_json = 2;
}

message StepResp {
  string observation_json = 1;
  double reward = 2;
  bool terminated = 3;
  bool truncated = 4;
  string info_json = 5;
}

message StepResultResp {
  string status = 1;   // submitted / success / pending / error
  optional string observation_json = 2;
  optional double reward = 3;
  optional bool terminated = 4;
  optional bool truncated = 5;
  optional string info_json = 6;
  optional string error = 7;
}

message CloseResp {
  bool success = 1;
}

message GetResetResultReq {
  string env_id = 1;
  int32 wait = 2;
}